proposals/0000-lazy-bundling.md

```diff
@@ -0,0 +1,126 @@
+
+## Drawbacks
+
+The key potential drawback is implementation complexity. However, the current proposal minimises this complexity by integrating lazy bundling into the core Metro algorithm (instead of creating a fork), and by keeping the integration surface between React Native and Metro as small as possible (encapsulated in `__loadBundleAsync` and `lazy=true`).
+
+## Alternatives
+
+This RFC is an evolution of an earlier implementation of lazy bundling at Meta, which:
+
+* Used a non-Fast-Refresh-compatible mechanism for passing bundle paths from the serializer to `asyncRequire`.
+* Did not include the `__loadBundleAsync` hook, instead closely coupling Metro to the React Native-specific way of fetching and evaluating bundles.
+* Treated lazy bundling as a transform-time config setting instead of a request parameter, resulting in worse caching the inability to build production and development bundles using the same Metro config.
+
+Apart from these implementation differences, the underlying concept remains the same as what we have been using at Meta since 2019, and we believe it is sound.
+
+## Adoption strategy / How we teach this
+
+Lazy bundling is not a breaking change from the perspective of React Native apps, so we can ship and enable it by default in the course of a single release of React Native. In development, all React Native apps already need to be connected to a working Metro server to support things like Fast Refresh, LogBox symbolication, and debugging - and this same requirement is sufficient for lazy bundling to work.
+
+From user code's perspective the API for lazy bundling is `import()` + `React.lazy()`. These are APIs that already work in Metro (and in React web apps in general) so they do not need to be taught, but they are probably not used very often in existing React Native code (since they do not perform any form of bundle splitting by default).
+
+To get the most benefit out of lazy bundling, developers will want to use a navigation library that uses `import()` as the mechanism for lazy-loading route components. We can work with leading libraries like React Navigation and Expo Router to move the ecosystem in this direction without requiring changes to user code.
+ 
+Optionally, authors of navigation libraries may find it useful to ship AST transformations (Babel plugins) that inject `import()` calls at *all* route boundaries in development, regardless of whether the associated screen is intended to be lazy-loaded in production. This would provide a consistent set of split points and opt users of those libraries into a more scalable developer experience.
+
+## Unresolved questions
+
+* User-facing naming: Are we happy with "lazy bundling" as the name for this feature?
+* APIs: Are we happy with `lazy=true` and `__loadBundleAsync` as the names for the new integration points between React Native and Metro?
```

Comment:
Lazy seems fine as it's reusing the naming from React, another option would be to use `async` as this is what we're doing with `require.context` (another Webpack concept).